### PR TITLE
Line memory profiler using memory hook and traceback

### PR DESCRIPTION
**Pull request by sonots**
This PR requires https://github.com/cupy/cupy/pull/264.

Code example::

    import cupy.cuda.memory_hooks
    hook = cupy.cuda.memory_hooks.LineProfileHook()
    with hook:
        # some cupy codes
    hook.print_report()

Output example::

    _root (4.00KB, 4.00KB)
      lib/python3.6/runpy.py:193:_run_module_as_main (4.00KB, 4.00KB)
        lib/python3.6/runpy.py:85:_run_code (4.00KB, 4.00KB)
          lib/python3.6/unittest/__main__.py:18:<module> (4.00KB, 4.00KB)
            lib/python3.6/unittest/main.py:94:__init__ (4.00KB, 4.00KB)
              lib/python3.6/unittest/main.py:255:runTests (4.00KB, 4.00KB)
                tests/cupy_tests/cuda_tests/memory_hooks_tests/test_line_profile.py:37:test_print_report (1.00KB, 1.00KB)
                tests/cupy_tests/cuda_tests/memory_hooks_tests/test_line_profile.py:38:test_print_report (1.00KB, 1.00KB)
                tests/cupy_tests/cuda_tests/memory_hooks_tests/test_line_profile.py:39:test_print_report (2.00KB, 2.00KB)

where the left and right in the parenthesis is *used bytes* and
*acquired bytes* respectively.
*Used bytes* is the memory bytes used from cupy memory pool, and
*Acquired bytes* is the actual memory bytes the cupy memory
pool acquired from GPU device.

### CODE REVIEW

**Comment by okuta** [comment]
Cloud you rebase this PR?

**Comment by sonots** [comment]
rebased to master

**Comment by okuta** [comment]
LGTM.
Cloud you add this to document?

**Comment by sonots** [comment]
Thank you for review. I will do it.

File: cupy/cuda/memory_hooks/line_profile.py [hunk 1]
```diff
@@ -0,0 +1,171 @@
+from os import path
+import sys
+import traceback
+
+from cupy.cuda.memory_hook import MemoryHook
+
+
+class LineProfileHook(MemoryHook):
+    """Code line Cupy memory profiler.
```
Comment (okuta) [inline cupy/cuda/memory_hooks/line_profile.py]: `Cupy` -> `CuPy`

File: cupy/cuda/memory_hooks/line_profile.py [hunk 2]
```diff
@@ -0,0 +1,171 @@
...
+
+    This profiler shows line-by-line memory consumption using traceback module.
+    But, note that it can trace only cpython level, no cython level.
```
Comment (okuta) [inline cupy/cuda/memory_hooks/line_profile.py]: `cpython` -> `CPython`
`cython` -> `Cython`

File: cupy/cuda/memory_hooks/line_profile.py [hunk 3]
```diff
@@ -0,0 +1,171 @@
...
+    ref. https://github.com/cython/cython/issues/1755
+
+    Example:
+        Code example::
+
+            from cupy.cuda import memory_hooks
+            hook = memory_hooks.LineProfileHook()
+            with hook:
+                # some cupy codes
```
Comment (okuta) [inline cupy/cuda/memory_hooks/line_profile.py]: `cupy` -> `CuPy`

File: cupy/cuda/memory_hooks/line_profile.py [hunk 4]
```diff
@@ -0,0 +1,171 @@
...
+                lib/python3.6/unittest/main.py:255:runTests (4.00KB, 4.00KB)
+                  tests/cupy_tests/test.py:37:test (1.00KB, 1.00KB)
+                  tests/cupy_tests/test.py:38:test (1.00KB, 1.00KB)
+                  tests/cupy_tests/test.py:39:test (2.00KB, 2.00KB)
+
+        Each line shows::
+
+            {filename}:{lineno}:{func_name} ({used_bytes}, {acquired_bytes})
+
+        where *used_bytes* is the memory bytes used from cupy memory pool, and
```
Comment (okuta) [inline cupy/cuda/memory_hooks/line_profile.py]: `cupy` -> `CuPy`

File: cupy/cuda/memory_hooks/line_profile.py [hunk 5]
```diff
@@ -0,0 +1,171 @@
...
+        *acquired_bytes* is the actual memory bytes the cupy memory pool
```
Comment (okuta) [inline cupy/cuda/memory_hooks/line_profile.py]: `cupy` -> `CuPy`

File: cupy/cuda/memory_hooks/line_profile.py [hunk 6]
```diff
@@ -0,0 +1,171 @@
...
+            self.lineno = obj[1]
+            self.name = obj[2]
+        else:  # >= 3.5 FrameSummary
+            self.filename = obj.filename
+            self.lineno = obj.lineno
+            self.name = obj.name
+
+
+class MemoryFrame(object):
+    """A single stack frame along with sum of memory usage at the frame
```
Comment (okuta) [inline cupy/cuda/memory_hooks/line_profile.py]: Please add period.

**Comment by sonots** [comment]
Fixed

File: cupy/cuda/memory_hooks/line_profile.py [hunk 7]
```diff
@@ -0,0 +1,171 @@
+from os import path
+import sys
+import traceback
+
+from cupy.cuda.memory_hook import MemoryHook
```
Comment (okuta) [inline cupy/cuda/memory_hooks/line_profile.py]: Sorry, I more point.
Please do not use direct class import.
Use `from cupy.cuda import memory_hook`

Comment (sonots) [reply cupy/cuda/memory_hooks/line_profile.py]: fixed

**Comment by okuta** [comment]
Thanks

**Pull request merged**